specificatie/datum.yaml

```diff
@@ -0,0 +1,119 @@
+openapi: 3.0.3
+info:
+  title: Datum definities
+  description: |
+    Definities voor de volgende datum types:
+    - datum conform LO GBA
+    - onbekend datum
+    - 'volledig' datum conform ISO-8601
+    - datum waarvan de dag onbekend is
+    - datum waarvan de maand en dag onbekend is 
+  version: 2.0.0
+  contact: {}
+paths: {}
+components:
+  schemas:
+    AbstractDatum:
+      type: object
+      required:
+        - type
+      properties:
+        type:
+          type: string
+      discriminator:
+        propertyName: type
+        mapping:
+          GbaDatum: '#/components/schemas/GbaDatum'
+          Datum: '#/components/schemas/VolledigDatum'
+          OnbekendDatum: '#/components/schemas/OnbekendDatum'
+          JaarDatum: '#/components/schemas/JaarDatum'
+          JaarMaandDatum: '#/components/schemas/JaarMaandDatum'
+    GbaDatumBasis:
+      type: object
+      properties:
+        datum:
+          type: string
+          pattern: ^[0-9]{8}$
+    GbaDatum:
+      description: Datum conform LO GBA
+      allOf:
+        - $ref: '#/components/schemas/AbstractDatum'
+        - $ref: '#/components/schemas/GbaDatumBasis'
+      required:
```

Comment:
Bovenstaande lijkt in conflict met [DD5.22](https://github.com/VNG-Realisatie/Haal-Centraal-common/blob/master/docs/design_decisions.md#dd522-bij-het-gebruik-van-allof-is-er-slechts-1-component-waarnaar-gerefereerd-wordt). Is dit toch niet correct of wordt het tijd DD5.22 aan te passen?

Reply:
DD5.22 geldt nog steeds, maar alleen in de genereerversie. @MelvLee ik dacht dat je had aangegeven dat dit bij genereren opgelost zou worden? Ik zie dit nog in de genereervariant terugkomen:
 ```
   GbaDatum:
      required:
      - datum
      description: Datum conform LO GBA
      example:
        value:
          type: GbaDatum
          datum: "20180700"
      allOf:
      - $ref: '#/components/schemas/AbstractDatum'
      - $ref: '#/components/schemas/GbaDatumBasis'
```

Reply:
Ik gaf aan dat code generatie correct genereert bij een allOf met 2 $ref's, niet dat het resolven er wat anders van maakt (bijv. expanden). Dit zal het resolven ook niet doen omdat dit valide OAS is.

Maar wat is het precieze probleem met deze constructie? Het is wat mij betreft geen probleem om dit als volgt uit te schrijven

```
    GbaDatum:
      description: Datum conform LO GBA
      allOf:
        - $ref: '#/components/schemas/AbstractDatum'
        - type: object
          properties:
            datum:
               type: string
               pattern: ^[0-9]{8}$
```

Ik vraag mij namelijk af wat het verschil is tussen het inline definieren van de tweede type of als een aparte type en die middels $ref te refereren. In beide gevallen inherit je van twee types.

Reply:
> Maar wat is het precieze probleem met deze constructie?

In het verleden hadden we (van jou?) begrepen dat deze constructie geen goede code zou genereren, omdat een object maar van één ander object kan overerven. Maar blijkbaar is er hier geen probleem.

Mag een allOf met 2 $refs dus altijd en is DD5.22 dus niet meer nodig, of kan dit alleen soms, zoals wanneer het eerste type een abstract type is?

Reply:
> > Maar wat is het precieze probleem met deze constructie?
> 
> In het verleden hadden we (van jou?) begrepen dat deze constructie geen goede code zou genereren, omdat een object maar van één ander object kan overerven. Maar blijkbaar is er hier geen probleem.
>
Dat klopt. Destijds werkte deze constructie, maar nu in dit geval wel.
 
> Mag een allOf met 2 $refs dus altijd en is DD5.22 dus niet meer nodig, of kan dit alleen soms, zoals wanneer het eerste type een abstract type is?

Een allOf met 2 $refs zal niet altijd tot de gewenste resultaat in code generatie leiden. Wat ik zie bij code generatie is dat de eerste $ref type wordt overgeërfd en de properties van de tweede $ref type wordt toegevoegd aan de kind type. Dus in het geval van GbaDatum ziet de gegenereerde code er als volgt uit:

```
class GbaDatum : AbstractDatum ( <= overerving van eerste $ref)
{
   string datum; ( <= overgenomen van de tweeder $ref)
}
```

Een allOf met 3 $ref's werkt zeker niet, dus je zou kunnen zeggen dat je DD5.22 kan versoepelen en zeggen dat meer dan 2 $refs niet is toegestaan. Maar zo simpel is het niet, want wat bij 2 $ref's niet tot het gewenste resultaat leidt, is als de type van de tweede $ref ook inherit van een andere type. De properties van deze derde type worden niet overgenomen door de codegenerator.

Ik weet dus niet of het handig is om hiervoor een DD op te nemen, want met deze rule beperk je de mogelijkheid om gedeelde properties te delen middels een afgeleide type